doc/src/devdocs/types.md

```diff
@@ -347,62 +234,54 @@ MyType{Int64,2}
 julia> MyType{Float32, 5}
 MyType{Float32,5}
 
-julia> MyType.name.cache
+julia> MyType.body.body.name.cache
 svec(MyType{Float32,5},MyType{Int64,2},#undef,#undef,#undef,#undef,#undef,#undef)
 ```
 
 (The cache is pre-allocated to have length 8, but only the first two entries are populated.) Consequently,
-when you instantiate a parametric type, each concrete type gets saved in a type-cache.  However,
-instances with `TypeVar` parameters are not cached.
+when you instantiate a parametric type, each concrete type gets saved in a type cache.  However,
+instances containing free type variables are not cached.
 
-## Tuple-types
+## Tuple types
 
-Tuple-types constitute an interesting special case.  For dispatch to work on declarations like
-`x::Tuple`, the type has to be able to be able to accommodate any tuple.  Let's check the parameters:
+Tuple types constitute an interesting special case.  For dispatch to work on declarations like
+`x::Tuple`, the type has to be able to accommodate any tuple.  Let's check the parameters:
 
 ```julia
 julia> Tuple
 Tuple
 
 julia> Tuple.parameters
-svec(Vararg{Any,N})
+svec(Vararg{Any,N} where N)
 ```
 
-It's worth noting that the parameter is a type, `Any`, rather than a `TypeVar T<:Any`: compare
-
-```julia
-julia> jl_(Tuple.parameters)
-svec(Vararg{Any, N<:Any})
-
-julia> jl_(Array.parameters)
-svec(T<:Any, N<:Any)
-```
-
-Unlike other types, tuple-types are covariant in their parameters, so this definition permits
-`Tuple` to match any type of tuple.  This is therefore equivalent to having an unbound `TypeVar`
-but distinct from a bound `TypeVar`
+Unlike other types, tuple types are covariant in their parameters, so this definition permits
+`Tuple` to match any type of tuple:
 
 ```julia
 julia> typeintersect(Tuple, Tuple{Int,Float64})
 Tuple{Int64,Float64}
 
 julia> typeintersect(Tuple{Vararg{Any}}, Tuple{Int,Float64})
 Tuple{Int64,Float64}
+```
 
-julia> T = TypeVar(:T,false)
-T
+However, if a variadic (`Vararg`) tuple type has free variables it can describe different kinds
+of tuples:
 
-julia> typeintersect(Tuple{Vararg{T}}, Tuple{Int,Float64})
+```julia
+julia> typeintersect(Tuple{Vararg{T} where T}, Tuple{Int,Float64})
 Tuple{Int64,Float64}
 
-julia> T = TypeVar(:T,true)
-T
-
-julia> typeintersect(Tuple{Vararg{T}}, Tuple{Int,Float64})
+julia> typeintersect(Tuple{Vararg{T}} where T, Tuple{Int,Float64})
 Union{}
 ```
 
-Finally, it's worth noting that `Tuple{}` is distinct
+Notice that when `T` is free with respect to the `Tuple` type (i.e. its binding `UnionAll`
+type is outside the `Tuple` type), only one `T` value must work over the whole type.
+Therefore a heterogeneous tuple does not match.
```

Comment:
I might've found the following flow easier to follow:

Unlike other types, tuple types are covariant in their parameters, so this definition permits
 `Tuple` to match any type of tuple:

```julia
julia> typeintersect(Tuple, Tuple{Int,Float64})
Tuple{Int64,Float64}

julia> typeintersect(Tuple{Vararg{Any}}, Tuple{Int,Float64})
Tuple{Int64,Float64}

julia> typeintersect(Tuple{Vararg{T} where T}, Tuple{Int,Float64})
Tuple{Int64,Float64}
```
However, if a variadic (`Vararg`) tuple type has free variables, it can describe other kinds
of tuples:
```julia
julia> typeintersect(Tuple{Vararg{T}} where T, Tuple{Int,Float64})
Union{}
```
Notice that when `T` is free with respect to the `Tuple` type (i.e. its binding `UnionAll`
type is outside the `Tuple` type), only one `T` value must work over the whole type.
Therefore a heterogeneous tuple does not match.